2020/3/trees/src/lib.rs

```diff
@@ -0,0 +1,53 @@
+use std::fs::File;
+use std::io::{self, BufRead};
+
+pub struct Map {
+    trees: Vec<String>
+}
+
+impl Map {
+    pub fn parse_file(path: &str) -> Map {
+        let file = File::open(path);
+    
+        let reader = io::BufReader::new(file.unwrap());
+        let mut trees = Vec::new();
+        for line in reader.lines() {
+            trees.push(line.unwrap());
+        }
+
+        Map::parse_lines(trees)
+    }
+
+    pub fn parse_string(map: &str) -> Map {
+        let mut trees = Vec::new();
+        for line in map.split_whitespace() {
+            trees.push(String::from(line));
+        }
+        Map::parse_lines(trees)
+    }
+
+    fn parse_lines(map: Vec<String>) -> Map {
+        Map { trees: map }
+    }
+
+    pub fn is_tree(&self, row: usize, col: usize) -> bool {
+        let len = self.trees[0].len();
+        let c = col % len;
+        self.trees[row].chars().nth(c).unwrap() == '#'
```

Comment:
The `nth` method is equivalent to `ElementAt` in LINQ. It's iterating the sequence in order to find the nth item. So it's fairly inefficient here.

Rust is hard-core about ensuring safety when working with its String (which are UTF-8 encoded). Since chars have a variable length, you can't index them, you can only enumerate them. So this is a fairly inefficient way to get the "nth" value from the map.

In my Go solution I loaded the map into a flag vector of bools (and tracked the "stride" in a separate field as well), which then let me index into it without iterating the body of the line every time.

That way you'd only enumerate the String once, to build the fixed-size data structure which you can then index into.

Reply:
Fixed in 231b8327a9bd50e66ea1a7abb7024bd1a7d4da9b